flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java

```diff
@@ -15,75 +16,468 @@
+						break;
+					case 2:
+						Entry<GlobalStreamId, DataStream<Tuple>> secondInput = iterator.next();
+						GlobalStreamId streamId2 = secondInput.getKey();
+						DataStream<Tuple> inputStream2 = secondInput.getValue();
+						outputStream = createOutput(boltId, userBolt, streamId, inputStream, streamId2, inputStream2);
+						break;
+					default:
+						throw new UnsupportedOperationException("Don't know how to translate a bolt "
+								+ boltId + " with " + numberOfInputs + " inputs.");
+				}
+
+				if (common.is_set_parallelism_hint()) {
+					int dop = common.get_parallelism_hint();
+					outputStream.setParallelism(dop);
+				} else {
+					common.set_parallelism_hint(1);
+				}
+
+			}
+		}
 	}
 
+	private DataStream<Tuple> processInput(String boltId, IRichBolt userBolt,
+										GlobalStreamId streamId, Grouping grouping,
+										Map<String, DataStream<Tuple>> producer) {
+
+		Preconditions.checkNotNull(userBolt);
+		Preconditions.checkNotNull(boltId);
+		Preconditions.checkNotNull(streamId);
+		Preconditions.checkNotNull(grouping);
+		Preconditions.checkNotNull(producer);
+
+		final String producerId = streamId.get_componentId();
+		final String inputStreamId = streamId.get_streamId();
+
+		DataStream<Tuple> inputStream = producer.get(inputStreamId);
+
+		final FlinkOutputFieldsDeclarer declarer = new FlinkOutputFieldsDeclarer();
+		declarers.put(boltId, declarer);
+		userBolt.declareOutputFields(declarer);
+		this.outputStreams.put(boltId, declarer.outputStreams);
+
+		// if producer was processed already
+		if (grouping.is_set_shuffle()) {
+			// Storm uses a round-robin shuffle strategy
+			inputStream = inputStream.rebalance();
+		} else if (grouping.is_set_fields()) {
+			// global grouping is emulated in Storm via an empty fields grouping list
+			final List<String> fields = grouping.get_fields();
+			if (fields.size() > 0) {
+				FlinkOutputFieldsDeclarer prodDeclarer = this.declarers.get(producerId);
+				inputStream = inputStream.keyBy(prodDeclarer
+						.getGroupingFieldIndexes(inputStreamId,
+								grouping.get_fields()));
+			} else {
+				inputStream = inputStream.global();
+			}
+		} else if (grouping.is_set_all()) {
+			inputStream = inputStream.broadcast();
+		} else if (!grouping.is_set_local_or_shuffle()) {
+			throw new UnsupportedOperationException(
+					"Flink only supports (local-or-)shuffle, fields, all, and global grouping");
+		}
+
+		return inputStream;
+	}
+
+	private SingleOutputStreamOperator<?, ?> createOutput(String boltId, IRichBolt bolt, GlobalStreamId streamId, DataStream<Tuple> inputStream) {
+		return createOutput(boltId, bolt, streamId, inputStream, null, null);
+	}
+
+	private SingleOutputStreamOperator<?, ?> createOutput(String boltId, IRichBolt bolt,
+														GlobalStreamId streamId, DataStream<Tuple> inputStream,
+														GlobalStreamId streamId2, DataStream<Tuple> inputStream2) {
+		Preconditions.checkNotNull(boltId);
+		Preconditions.checkNotNull(streamId);
```

Comment:
Asserts are not enabled by default. I think it's important to always check this condition.